brave/src/main/java/com/linecorp/armeria/client/brave/BraveClient.java

```diff
@@ -68,81 +69,90 @@
-        tracer = tracing.tracer();
-        injector = tracing.propagationFactory().create(AsciiStringKeyFactory.INSTANCE)
-                          .injector(RequestHeadersBuilder::set);
-        this.remoteServiceName = remoteServiceName;
+        currentTraceContext = httpTracing.tracing().currentTraceContext();
+        tracer = httpTracing.tracing().tracer();
+        clientParser = httpTracing.clientParser();
+        adapter = new ArmeriaHttpClientAdapter();
+        handler = HttpClientHandler.create(httpTracing, adapter);
+        injector = httpTracing.tracing().propagationFactory().create(AsciiStringKeyFactory.INSTANCE)
+                              .injector(RequestHeadersBuilder::set);
     }
 
     @Override
     public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Exception {
-        final Span span = tracer.nextSpan();
-
-        // Inject the headers.
         final RequestHeadersBuilder newHeaders = req.headers().toBuilder();
-        injector.inject(span.context(), newHeaders);
+        final Span span = handler.handleSend(injector, newHeaders, ctx.log());
         req = HttpRequest.of(req, newHeaders.build());
         ctx.updateRequest(req);
 
+        // Ensure the trace context propagates to children
+        ctx.onChild(TraceContextUtil::copy);
+
         // For no-op spans, we only need to inject into headers and don't set any other attributes.
         if (span.isNoop()) {
-            return delegate().execute(ctx, req);
+            try (SpanInScope ignored = tracer.withSpanInScope(span)) {
+                return delegate().execute(ctx, req);
+            }
         }
 
-        final String method = ctx.method().name();
-        span.kind(Kind.CLIENT).name(method);
         ctx.log().addListener(log -> SpanContextUtil.startSpan(span, log),
                               RequestLogAvailability.REQUEST_START);
 
-        // Ensure the trace context propagates to children
-        ctx.onChild(TraceContextUtil::copy);
-
         ctx.log().addListener(log -> {
             SpanTags.logWireSend(span, log.requestFirstBytesTransferredTimeNanos(), log);
-
             // If the client timed-out the request, we will have never received any response data at all.
             if (log.isAvailable(RequestLogAvailability.RESPONSE_FIRST_BYTES_TRANSFERRED)) {
                 SpanTags.logWireReceive(span, log.responseFirstBytesTransferredTimeNanos(), log);
             }
-
-            finishSpan(span, log);
+            setRemoteEndpoint(span, log);
+            handleFinish(log, span);
         }, RequestLogAvailability.COMPLETE);
 
         try (SpanInScope ignored = tracer.withSpanInScope(span)) {
             return delegate().execute(ctx, req);
         }
     }
 
-    private void finishSpan(Span span, RequestLog log) {
-        setRemoteEndpoint(span, log);
-        SpanContextUtil.closeSpan(span, log);
-    }
-
     private void setRemoteEndpoint(Span span, RequestLog log) {
```

Comment:
sorry I told you wrong, if you implement `adapter.parseClientIpAndPort(request, span)` it is done automatically

Reply:
you mean parseServerIpAndPort?

Reply:
no I mean implement in the client adapter class parseClientIpAndPort() then you do nothing here. if unclear I can contribute to this pull request if you grant me access to your fork

Reply:
oops yeah this is client! sorry I forget both client and server are here. you are right to do it explicitly. my mistake

Reply:
ah you mean server side will handle this.